### PR TITLE
[#2932] Skip KVK branch selection if vestigingsnummer already selected

### PR DESCRIPTION
**Pull request by pi-sigma**
The KVK branch selection page should be skipped for users who have already selected a branch on a different platform

- [x] Update eHerkenning OIDC flow: get vestigingsnummer from OIDC claim (if present) and store in session
- [x] Update eHerkenning SAML flow: get vestigingsnummer from SAML attributes (if present) and store in session
- [x] Skip KVK branch selection if vestigingsnummer already in session

Taiga: https://taiga.maykinmedia.nl/project/open-inwoner/task/2932

### CODE REVIEW

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: I think this is very much worth adding to the library, especially if the PR Is small (we can keep it here for now to test our flow, but it would be good to ensure we upstream our changes and eventually revert our override when the library is fixed).

**Comment by swrichards** [comment]
Braindump: @pi-sigma this is technically orthogonal to the issue here (skipping the branch if already selected during the eHerkenning flow), but it's also important that users can see _only_ vestigingen they're authorized to see according to the eHerkenning scope. I am not sure if that flows from this claim (Selected branch) versus some other claim that may or may not exist (all authorized branches).

**Comment by pi-sigma** [comment]
@swrichards Good point. Is this already covered by our use of `get_kvk_branch_number` to scope the retrieval of roles, zaken, and vragen to a particular vestiging, or are we missing something?

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [reply src/open_inwoner/accounts/backends.py]: As discussed off-line, we'll try something that requires a bit less overloading of third party libraries:

```python
def authenticate(self, *args, **kwargs):
    result = super().authenticate(*args, **kwargs)
    if result is not None:
        # TODO: Ensure the setting to persist the ID token in the session is set
        id_token = self.request.session["oidc_id_token"]
        user_info = self.verify_token(id_token)
        if vestigingsnummer := user_info.get("vestigingsnummer", None):
            self.request.session[KVK_BRANCH_SESSION_VARIABLE] = vestigingsnummer
            self.request.session.save()
    
    return result
```

Comment (swrichards) [reply src/open_inwoner/accounts/backends.py]: Alternative no. 2, nicer because it stays closer to the public API:

```python
# Misusing the public interface to gain access to claims, which are not
# otherwise exposed but are needed to access the vestigingsnummer.
def create_user(self, claims):
    user = super().create_user(claims)
    if vestigingsnummer := claims.get("vestigingsnummer", None):
        self.request.session[KVK_BRANCH_SESSION_VARIABLE] = vestigingsnummer
        self.request.session.save()

    return user

def update_user(self, user, claims):
    user = super().update_user(claims)
    if vestigingsnummer := claims.get("vestigingsnummer", None):
        self.request.session[KVK_BRANCH_SESSION_VARIABLE] = vestigingsnummer
        self.request.session.save()
    
    return user
```

File: src/open_inwoner/accounts/tests/test_oidc_views.py
Comment (swrichards) [inline src/open_inwoner/accounts/tests/test_oidc_views.py:1851]: We should make sure that your overrides are actually hit: currently, I don't think the bit of code that actually sets the vestigingsnummer in the session is being tested.

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: Note the helpers we have for this in `open_inwoner.kvk.brances`.

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: Shouldn't we still call the `super().update_user`?

**Review by swrichards — CHANGES_REQUESTED** [review]

File: src/open_inwoner/accounts/tests/test_oidc_views.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/tests/test_oidc_views.py:1851]: Updated the test

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: Went for #2

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: As discussed, we don't have a helper function for storing the value in the session, and since the functions are quite simple, adding another one doesn't seem to be of much use. Let's continue importing and using `KVK_BRANCH_SESSION_VARIABLE`

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: The parent class in in `digid-eherkenning/oidc/backends` overrides the `update_user` from `mozilla_django_oidc_db` method to do nothing, so I though there's no point in calling `super`. I added it back in, though.

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: I am still somewhat worried by the fact that `branch_number_claim` is a list. Can we assume it always has 1 item? If not, should we iterate over the list and return the first vestigingsnummer found to be future-proof?

**Review by swrichards — APPROVED** [review]
Just one question about the claims: I'll leave it up to you whether we need to fix this now or whether it needs fixing at all.

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: Fair enough, I don't have any experience with the format of these claims either. Will change as suggested.

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: Should this be in this method? It's strictly speaking about the vestigingsnummer. If we're logging anything here, I'd say we log that we did (or did not) find a vestigingsnummer, and leave the "we created or updated a user" in the function where that happens.

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: Just out of interest: did we figure out if there are ever multiple claims? No need to hold up the PR for that, just curious.

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: Are we fetching this only for the logging? In that case I'd also say we log this earlier in the chain, there's a lot of escape hatches in this function making it slightly harder to follow, if we can simplify it that'd be better.

File: src/open_inwoner/accounts/backends.py
Comment (swrichards) [inline src/open_inwoner/accounts/backends.py]: My sense is this is an optional in a "you might not have eHerkenning set up" kind of way, but if we make it to this function, eHerkenning is evidently configured, and then I feel a Sentry ping would be in order to at least alert us to the incomplete configuration.

**Review by swrichards — CHANGES_REQUESTED** [review]
Very small nitpicks, otherwise GTG!

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: You're right, that "retrieve/create" dichotomy indicates that the message is rather specific to the calling functions. I'll change and simplify accordingly.

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: No. I'm not sure what the use case would be, but the library definitely makes room for this.

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: I think the escape hatch for the `identifier_type_claim` is unnecessary. If you're logging in with eHerkenning, the KVK or RSIN needs to be in the claims

File: src/open_inwoner/accounts/backends.py
Comment (pi-sigma) [reply src/open_inwoner/accounts/backends.py]: In fact, the `_check_candidate_backend` already takes care of this, and since the config class provides a default for the claim field, this is guaranteed to exist. Will simplify

**Review by swrichards — APPROVED** [review]

**Review by alextreme — APPROVED** [review]

**Pull request merged**
